### PR TITLE
[System.Drawing.Common] Work around libgdiplus use after free

### PR DESCRIPTION
**Pull request by lambdageek**
On Windows, both of the following are legal

    Metafile mf = ... ; // get a metafile instance
    Graphics g = Graphics.FromImage(mf);
    g.Dispose(); mf.Dispose();

and

    Metafile mf = ... ; // get a metafile instance
    Graphics g = Graphics.FromImage(mf);
    mf.Dispose(); g.Dispose();

On Unix, libgdiplus has a use after free bug for the second form - the metafile
native image is disposed, but the graphics instance still has a pointer to the
memory that it will use during cleanup.  If the memory is reused, the graphics
instance will see garbage values and crash.

The workaround is to add a MetadataHolder class and to transfer responsibility
for disposing of the native image instance to it if the Metafile is disposed
before the Graphics.

Note that the following is not allowed (throws OutOfMemoryException on GDI+ on
Windows), so there's only ever one instance of Graphics associated with a
Metafile at a time.

    Graphics g = Graphics.FromImage(mf);
    Graphics g2 = Graphics.FromImage(mf); // throws

Addresses https://github.com/dotnet/runtime/issues/37838

### CODE REVIEW

**Comment by lambdageek** [comment]
One thing I'm not sure about - is it worthwhile to make `Metafile.MetafileHolder` a subclass of `SafeHandle`.  On construction I'd bump the refcount to 2 but leave the handle as invalid, then MetafileDispose would call `SetHandle` and `SafeHandle.Dispose`, and GraphicsDispose would also call `SafeHandle.Dispose`.

One reason not to do it might be because the rest of the SafeHandle methods aren't really meaningful to `MetafileHolder` - it has quite a constrained API surface.  I wish `SafeHandle` was a concrete class with a delegate for the release method, then I could just use it as a field in MetafileHolder.

File: src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs [hunk 1]
```diff
@@ -43,6 +44,95 @@ namespace System.Drawing.Imaging
...
+
+            internal void Dispose(bool disposing)
+            {
+                if (disposed)
+                    return;
+                IntPtr nativeImage = _nativeImage;
+                _nativeImage = IntPtr.Zero;
+                disposed = true;
+                if (nativeImage != IntPtr.Zero)
+                    Metafile.DisposeNativeInstance (nativeImage);
```
Comment (safern) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs:106]:
```suggestion
                if (!disposed)
                {
                    IntPtr nativeImage = _nativeImage;
                    _nativeImage = IntPtr.Zero;
                    disposed = true;
                    if (nativeImage != IntPtr.Zero)
                        Metafile.DisposeNativeInstance (nativeImage);
                 }
```

File: src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs [hunk 2]
```diff
@@ -43,6 +44,95 @@ namespace System.Drawing.Imaging
...
+
+
+            internal bool Disposed { get => disposed; }
+            internal MetafileHolder ()
+            {
+                disposed = false;
+                _nativeImage = IntPtr.Zero;
+            }
+
+            ~MetafileHolder () => Dispose(false);
```
Comment (safern) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs:90]:
```suggestion
            ~MetafileHolder() => Dispose(false);
```

File: src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs [hunk 3]
```diff
@@ -43,6 +44,95 @@ namespace System.Drawing.Imaging
...
+        //      Graphics g = Graphics.FromImage(mf);
+        //      Graphics g2 = Graphics.FromImage(mf);  // throws OutOfMemoryException on GDI+ on Win32
+        internal sealed class MetafileHolder : IDisposable
+        {
+            private bool disposed;
+            private IntPtr _nativeImage;
+
+
+            internal bool Disposed { get => disposed; }
+            internal MetafileHolder ()
```
Comment (safern) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs:84]:
```suggestion
            internal MetafileHolder()
```

File: src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs [hunk 4]
```diff
@@ -43,6 +44,95 @@ namespace System.Drawing.Imaging
...
+        // instance is disposed.
+        //
+        // Note that the following throws, so we only ever need to keep track of one Graphics
+        // instance at a time:
+        //      Metafile mf = ...;  // get a metafile instance
+        //      Graphics g = Graphics.FromImage(mf);
+        //      Graphics g2 = Graphics.FromImage(mf);  // throws OutOfMemoryException on GDI+ on Win32
+        internal sealed class MetafileHolder : IDisposable
+        {
+            private bool disposed;
```
Comment (safern) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs:79]:
```suggestion
            private bool _disposed;
```

File: src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs [hunk 5]
```diff
@@ -43,6 +44,95 @@ namespace System.Drawing.Imaging
...
+                if (disposed)
+                    return;
+                IntPtr nativeImage = _nativeImage;
+                _nativeImage = IntPtr.Zero;
+                disposed = true;
+                if (nativeImage != IntPtr.Zero)
+                    Metafile.DisposeNativeInstance (nativeImage);
+            }
+
+            internal void MetafileDisposed (IntPtr nativeImage)
```
Comment (safern) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs:109]:
```suggestion
            internal void MetafileDisposed(IntPtr nativeImage)
```

File: src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs [hunk 6]
```diff
@@ -43,6 +44,95 @@ namespace System.Drawing.Imaging
...
+            {
+                _nativeImage = nativeImage;
+            }
+
+            internal void GraphicsDisposed ()
```
Comment (safern) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs:114]:
```suggestion
            internal void GraphicsDisposed()
```

File: src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs [hunk 7]
```diff
@@ -43,6 +44,95 @@ namespace System.Drawing.Imaging
...
+            {
+                Dispose();
+            }
+        }
+
+        internal MetafileHolder? AddMetafileHolder(Graphics g)
```
Comment (safern) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs:120]: Why do we need the `Graphics` here?

File: src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs [hunk 8]
```diff
@@ -43,6 +44,95 @@ namespace System.Drawing.Imaging
...
+        {
+            // If _metafileHolder is not null and hasn't been disposed yet, there's already a graphics instance associated with
+            // this metafile, the native code will return an error status.
+            if (_metafileHolder != null && !_metafileHolder.Disposed)
+                return null;
+            _metafileHolder = new MetafileHolder();
+            return _metafileHolder;
+        }
+
+        internal static void DisposeNativeInstance(IntPtr nativeImage)
```
Comment (safern) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs:130]: Why make this a method if it is only used in one place? Should we just put this inline?

File: src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs [hunk 9]
```diff
@@ -143,6 +233,20 @@ public Metafile(IntPtr hmetafile, WmfPlaceableFileHeader wmfHeader)
...
         }
 
+        protected override void Dispose(bool disposing)
+        {
+            if (_metafileHolder != null && !_metafileHolder.Disposed)
+            {
+                // There's a graphics instance created from this Metafile,
+                // transfer responsibility for disposing the nativeImage to the
+                // MetafileHolder
+                _metafileHolder.MetafileDisposed (nativeImage);
```
Comment (safern) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs:243]:
```suggestion
                _metafileHolder.MetafileDisposed(nativeImage);
```

File: src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs [hunk 10]
```diff
@@ -143,6 +233,20 @@ public Metafile(IntPtr hmetafile, WmfPlaceableFileHeader wmfHeader)
...
+                _metafileHolder = null;
+                nativeImage = IntPtr.Zero;
+            }
+            base.Dispose (disposing);
```
Comment (safern) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs:247]:
```suggestion
            base.Dispose(disposing);
```

File: src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs [hunk 11]
```diff
@@ -143,6 +233,20 @@ public Metafile(IntPtr hmetafile, WmfPlaceableFileHeader wmfHeader)
...
+        {
+            if (_metafileHolder != null && !_metafileHolder.Disposed)
+            {
+                // There's a graphics instance created from this Metafile,
+                // transfer responsibility for disposing the nativeImage to the
+                // MetafileHolder
+                _metafileHolder.MetafileDisposed (nativeImage);
+                _metafileHolder = null;
+                nativeImage = IntPtr.Zero;
+            }
```
Comment (safern) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs:246]: Add extra line.

**Comment by safern** [comment]
cc: @stephentoub

**Review by safern — APPROVED** [review]

File: src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs [hunk 12]
```diff
@@ -43,6 +44,93 @@ namespace System.Drawing.Imaging
...
+        // when the managed Metafile instance is disposed while a Graphics instance is still
+        // not disposed (ie the second code pattern above) and to keep the native image alive until the graphics
+        // instance is disposed.
+        //
+        // Note that the following throws, so we only ever need to keep track of one Graphics
+        // instance at a time:
+        //      Metafile mf = ...;  // get a metafile instance
+        //      Graphics g = Graphics.FromImage(mf);
+        //      Graphics g2 = Graphics.FromImage(mf);  // throws OutOfMemoryException on GDI+ on Win32
+        internal sealed class MetafileHolder : IDisposable
```
Comment (RussKie) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs:77]: Just for my own curiosity, any reason not to make it a struct?

Comment (ericstj) [reply src/libraries/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.Unix.cs:77]: Looks like it has a finalizer.  Structs can't have finalizers.

**Pull request merged**